packages/netlify-cms-backend-github/src/GraphQLAPI.js

```diff
@@ -0,0 +1,459 @@
+    const { repo_owner: owner, repo_name: name } = this;
+
+    return {
+      query: queries.pullRequestAndBranch,
+      variables: { owner, name, number, qualifiedName: this.getBranchQualifiedName(branch) },
+    };
+  }
+
+  async pullRequestAndBranch(branch, number) {
+    const { data } = await this.query({
+      ...this.getPullRequestAndBranchQuery(branch, number),
+      fetchPolicy: CACHE_FIRST,
+    });
+
+    return data.repository;
+  }
+
+  async deleteUnpublishedEntry(collection, slug) {
+    try {
+      const contentKey = slug;
+      const branchName = this.generateBranchName(contentKey);
+
+      const metadata = await this.retrieveMetadata(contentKey);
+      const { branch, pullRequest } = await this.pullRequestAndBranch(
+        branchName,
+        metadata.pr.number,
+      );
+
+      const { data } = await this.mutate({
+        mutation: mutations.closePullRequest,
+        variables: {
+          deleteRefInput: { refId: branch.id },
+          closePullRequestInput: { pullRequestId: pullRequest.id },
+        },
+        update: store => {
+          store.data.delete(defaultDataIdFromObject(branch));
+          store.data.delete(defaultDataIdFromObject(pullRequest));
+        },
+      });
+
+      return data.closePullRequest;
+    } catch (e) {
+      const { graphQLErrors } = e;
+      if (graphQLErrors && graphQLErrors.length > 0) {
+        const branchNotFound = graphQLErrors.some(e => e.type === 'NOT_FOUND');
+        if (branchNotFound) {
+          return;
+        }
+      }
+      throw e;
+    }
+  }
+
+  async createBranchAndPullRequest(branchName, sha, title) {
+    const { id, __typename } = await this.getRepository();
+    const { data } = await this.mutate({
+      mutation: mutations.createPullRequest,
+      variables: {
+        createRefInput: {
+          name: this.getBranchQualifiedName(branchName),
+          oid: sha,
+          repositoryId: id,
+        },
+        createPullRequestInput: {
+          baseRefName: this.branch,
+          body: 'Automatically generated by Netlify CMS',
+          title,
+          headRefName: branchName,
+          repositoryId: id,
+        },
+      },
+      update: (store, { data: mutationResult }) => {
+        const { branch } = mutationResult.createRef;
+        const { pullRequest } = mutationResult.createPullRequest;
+        const branchData = { repository: { id, branch, __typename } };
+        const pullRequestData = { repository: { id, branch, pullRequest, __typename } };
+
+        store.writeQuery({
+          ...this.getBranchQuery(branchName),
+          data: branchData,
+        });
+
+        store.writeQuery({
+          ...this.getPullRequestAndBranchQuery(branchName, pullRequest.number),
+          data: pullRequestData,
+        });
+      },
+    });
+    const { pullRequest } = data.createPullRequest;
+    return { ...pullRequest, head: { sha: pullRequest.headRefOid } };
+  }
+
+  async getTree(sha) {
```

Comment:
We can actually do a deep nested query to get the entire repo structure in one query, but since there is already some code in the base class to handle the recursion, it will require a more significant change.